Process user task update command

tmetzke opened the pull request:
## Description

* Introduces an UPDATE command processor and related appliers for UPDATING and UPDATED events.
* Refactors the precondition checker to receive additional checks externally.
* Further reasonings about implementation decisions can be found in #15530.

## Related issues

<!-- Which issues are closed by this PR or are related -->

closes #15530

<!-- Cut-off marker
_All lines under and including the cut-off marker will be removed from the merge commit message_

## Definition of Ready

Please check the items that apply, before requesting a review.

You can find more details about these items in our wiki page about [Pull Requests and Code Reviews](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews).

* [ ] I've reviewed my own code
* [ ] I've written a clear changelist description
* [ ] I've narrowly scoped my changes
* [ ] I've separated structural from behavioural changes
-->

## Definition of Done

<!-- Please check the items that apply, before merging or (if possible) before requesting a review. -->

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

ana-vinogradova-camunda commented on engine/src/main/java/io/camunda/zeebe/engine/processing/usertask/UserTaskClaimProcessor.java:
👍🏻 really like how the refactored code looks. Thank you for the idea to keep the command-specific checks in the command processors.

ana-vinogradova-camunda commented on engine/src/test/java/io/camunda/zeebe/engine/processing/usertask/UpdateUserTaskTest.java:
👍🏻 Such detailed test coverage, nice!

ana-vinogradova-camunda reviewed (approved):
👍🏻 LGTM, very well structured PR and commits, that makes it easy to review, nice!

tmetzke commented:
**Review note** (@koevskinikola and @ana-vinogradova-camunda)
I am currently not exporting/serializing the `changedAttributes` property of the user task records. Thus, when exported, the event records don't contain this data. There might be some value in having this information exported as well down the line. If you think we should export this already to avoid adjusting this later again, let me know. I simply wasn't sure whether this is a good idea already or not.

koevskinikola commented on engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/activity/NativeUserTaskTest.java:
```diff
@@ -608,6 +609,42 @@ public void shouldClaimUserTask() {
+    final long processInstanceKey = ENGINE.processInstance().ofBpmnProcessId(PROCESS_ID).create();
+
+    // when
+    ENGINE.userTask().ofInstance(processInstanceKey).update(new UserTaskRecord());
```
:thinking: I wonder if we should assert the initial values of the `UserTaskRecord`. This would allow us to detect any regressions when some of these properties are not set.

tmetzke replied:
I thought about it as well and wasn't sure if that's overkill. Since we both thought about it now, I have enough evidence to actually add it. 😄

koevskinikola reviewed (changes requested):
@tmetzke the code looks great!

I only have a minor change request regarding the new `UserTaskRecord` property.

Also, regarding your comment [here](https://github.com/camunda/zeebe/pull/15819#issuecomment-1895285047), I think it would be better if we export that property as well. AFAIK, Operate uses the `zeebe-protocol-impl` module when importing Zeebe data. I wonder if it will be an issue if some of the data is missing.

tmetzke commented on protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/usertask/UserTaskRecord.java:
```diff
@@ -51,6 +61,8 @@ public final class UserTaskRecord extends UnifiedRecordValue implements UserTask
+      new ArrayProperty<>("changedAttributes", StringValue::new);
 
   public UserTaskRecord() {
     super(15);
```
Nice catch, really good, thank you! 👍

tmetzke replied:
Fixed with https://github.com/camunda/zeebe/pull/15819/commits/ca5cdeeadd71f370afd328624fb45416546aa423

tmetzke commented on engine/src/main/java/io/camunda/zeebe/engine/processing/usertask/UserTaskCommandPreconditionChecker.java:
Thought about it but refrained from doing it here. It's a bit more complex than I initially thought since we have more commands already that have slightly different needs.

koevskinikola replied:
We have a separate issue for that one. I just mentioned it since how you did it provides a lot of flexibility to the code.

tmetzke replied:
Makes sense 👍

tmetzke commented:
@koevskinikola, thanks for the review feedback, really helpful as usual ⭐ 
I addressed your suggestions and serialize+export the new attribute now as well.

I worked around the issue with the state updates for now. If you see a better way of doing that, let me know. I refrained from copying the resulting record from the column family each time I do a `getUserTask` as I find that a bit too expensive. Let me know if you think otherwise.

koevskinikola reviewed (approved):
@tmetzke the code looks good now. :tada: 

I only have some feedback for the place where we clone the buffer of a user task record when we update it. I think we should move it to follow the existing pattern from other places.

But the current behavior is still correct, so I'm approving the PR.

tmetzke commented on engine/src/main/java/io/camunda/zeebe/engine/state/appliers/UserTaskUpdatedApplier.java:
```diff
@@ -0,0 +1,37 @@
+
+  @Override
+  public void applyState(final long key, final UserTaskRecord value) {
+    // prepare the changeset
+    final UserTaskRecord updateRecord = new UserTaskRecord();
+    updateRecord.wrapChangedAttributes(value, true);
```
Makes total sense, I also considered that option 👍 
I'll move this to the processor class 🙂

tmetzke replied:
I adjusted the UPDATE(D) processor and applier with https://github.com/camunda/zeebe/pull/15819/commits/2cf410ffa9a6abeee39763db3678254f84d7e937. I copied the persisted record and applied the changed attributes (including the tracking property) so all new values of the user task are visible in the events.

I did not change the ASSIGN(ED) processor and applier. I consider copying the record in the processor merely for the `assignee` property unjustified. Let me know if you disagree.
